proposals/lip-0051.md

```diff
@@ -1267,46 +1175,95 @@ unlock(address, moduleID, tokenID, amount):
         or lockedAmount(address, moduleID, tokenID) < amount):
         unlock fails
 
-    availableBalance(address, tokenID) += amount
     lockedAmount(address, moduleID, tokenID) -= amount
+    availableBalance(address, tokenID) += amount
+```
+
+#### forwardMessageFee
+
+This function is only called during the execution of a cross-chain update on the mainchain.
+```python
+forwardMessageFee(ccm):
+    fee = ccm.fee
+    feeTokenID = interoperability.getChannel(ccm.sendingChainID).messageFeeTokenID
+    if (fee < 0
+        or feeTokenID != interoperability.getChannel(ccm.receivingChainID).messageFeeTokenID):
+        forwardMessageFee fails
+
+    
+    if (feeTokenID.chainID != interoperability.getOwnChainAccount().ID
+        or escrowAmount(ccm.sendingChainID, feeTokenID.localID) < fee):
+        interoperability.terminateChain(ccm.sendingChainID)
+        return False
+
+    # transfer the fee from escrow to escrow
+    escrowAmount(ccm.sendingChainID, feeTokenID.localID) -= fee
+    escrowAmount(ccm.receivingChainID, feeTokenID.localID) += fee
 ```
 
-#### beforeExecuteCCM
+
+#### beforeApplyCCM
 
 ```python
-beforeExecuteCCM(relayerAddress, CCM):
-    fee = CCM.fee
+beforeApplyCCM(ccu, ccm):
+    relayerAddress = address derived from ccu.senderPublicKey
+    fee = ccm.fee
     if fee < 0:
-        beforeExecuteCCM fails
-
-    if interoperability.getOwnChainAccount().ID == MAINCHAIN_ID:
-        # If this chain is the mainchain, unescrow the fee
-        if escrowAmount(CCM.sendingChainID, LOCAL_ID_LSK) < fee:
-            beforeExecuteCCM fails
-        availableBalance(relayerAddress, TOKEN_ID_LSK_MAINCHAIN) += fee
-        escrowAmount(CCM.sendingChainID, LOCAL_ID_LSK) -= fee
+        beforeApplyCCM fails
+
+    feeTokenID = interoperability.getChannel(ccm.sendingChainID).messageFeeTokenID
+    # If the chain is the fee token native chain, unescrow and assign fee to relayer
+    if feeTokenID.chainID == interoperability.getOwnChainAccount().ID:
+        if escrowAmount(ccm.sendingChainID, feeTokenID.localID) < fee:
+            interoperability.terminateChain(ccm.sendingChainID)
+            return False
```

Comment:
In these API, we sometimes write `beforeApplyCCM fails` and sometimes return `False`. Are these two cases different, i.e., do we expect different error handling? To me `fail` would mean throwing an exception, i..e, not intended normal program flow, whereas `return False` is not failing, but simply a return value based on the expected program flow.

I think we should be consistent here and also it should be consistent with the error handling on the caller side. In particular, if in the interoperability module we write  
> if the previous call fails:
>            return False
Then we should fail here if we expect that logic to be triggered.

Reply:
I would check the occurrences of `fail` and `False` in the document, as I think they are mixed in other places as well.

Reply:
I think APIs should fail when they can't be resolved (and not return `False`). Iker pointed out that in this case it could still be better to return `False` as the API did not really fail to execute, but rather executed by terminating the chain. 
I think we don't have a clear error handling system in place, and we should probably discuss it before we change the LIP further in that regard.

Reply:
I think this point can be resolved as we handle it as part of the error handling changes.